torch/testing/_internal/logging_tensor.py

```diff
@@ -74,52 +94,100 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
-            o._shortid = self.next_shortid  # type: ignore[attr-defined]
-            self.next_shortid += 1
-        return o._shortid  # type: ignore[attr-defined]
-
-    def _fmt(self, a: object) -> str:
+        self.tracebacks_list = tracebacks_list
+        self.memo = {}
+        self.next_id = 0
+        self.with_type = with_type
+
+    def _shortid(self, t: torch.Tensor) -> int:
+        o = WeakIdRef(t)
+        weak_self = weakref.ref(self)
+
+        def del_memo():
+            self = weak_self()
+            if self is None:
+                return
+            self.memo.pop(o, None)
+
+        weakref.finalize(t, del_memo)
+        if o not in self.memo:
+            self.memo[o] = self.next_id
+            self.next_id += 1
+        return self.memo[o]
+
+    def _fmt(self, a: object, with_type: bool = False) -> str:
         cond_cls = torch.Tensor if self.use_shortid_for_all_tensors else LoggingTensor
-        return f'${self._shortid(a)}' if isinstance(a, cond_cls) else repr(a)
+        if isinstance(a, cond_cls):
+            maybe_type = ""
+            if with_type and self.with_type:
+                maybe_type = f": {_dtype_abbrs[a.dtype]}[{', '.join(map(str, a.shape))}]"
+            x = f"${self._shortid(a)}{maybe_type}"
+            return x
+        else:
+            return repr(a)
 
     def emit(self, record):
-        fmt_args = ", ".join(itertools.chain(
-            (self._fmt(a) for a in record.args[0]),
-            (f"{k}={self._fmt(v)}" for k, v in record.args[1].items())
-        ))
-        fmt_rets = ", ".join(self._fmt(a) for a in record.args[2]) \
-            if isinstance(record.args[2], (list, tuple)) else self._fmt(record.args[2])
+        fmt_args = ", ".join(
+            itertools.chain(
+                (_maybe_repr(tree_map(self._fmt, a)) for a in record.args[0]),
+                (f"{k}={_maybe_repr(tree_map(self._fmt, v))}" for k, v in record.args[1].items()),
+            )
+        )
+        fmt_rets = tree_map(functools.partial(self._fmt, with_type=True), record.args[2])
         self.log_list.append(f'{fmt_rets} = {record.msg}({fmt_args})')
+        if self.tracebacks_list is not None:
+            self.tracebacks_list.append(record.traceback)
 
 def log_input(name: str, var: object):
-    logging.getLogger("LoggingTensor").info("input", (name,), {}, (var,))
+    logging.getLogger("LoggingTensor").info("input", (name,), {}, var)
+
+class GatherTraceback(logging.Filter):
+    def __init__(self, python=True, script=True, cpp=False):
+        self.python = python
+        self.script = script
+        self.cpp = cpp
+
+    def filter(self, record):
+        record.traceback = gather_traceback(python=self.python, script=self.script, cpp=self.cpp)
+        return True
 
 @contextlib.contextmanager
-def capture_logs(is_mode=False) -> Iterator[List[str]]:
+def capture_logs(is_mode=False, python_tb=False, script_tb=False, cpp_tb=False) -> Iterator[List[str]]:
```

Comment:
I think the Mode version is the one I usually use, it would be great to add these args to capture_logs_with_logging_tensor_mode below